WordPress/src/main/java/org/wordpress/android/ui/accounts/login/nav/LoginNavController.java

```diff
@@ -0,0 +1,150 @@
+            return new InputSiteAddressHandler();
+        }
+
+        throw new RuntimeException("Unsupported login state " + loginNav.getSimpleName());
+    }
+
+    private Stack<Class<? extends LoginNav>> mLoginNavStack = new Stack<>();
+
+    private boolean isInState(Class<? extends LoginNav> loginNav) {
+        return !mLoginNavStack.empty() && loginNav.equals(mLoginNavStack.peek());
+    }
+
+    public void ensureState(Class<? extends LoginNav> loginNav) {
+        if (!isInState(loginNav)) {
+            throw new RuntimeException("Not in state " + loginNav.getSimpleName());
+        }
+    }
+
+    private void gotoState(Class<? extends LoginNav> loginNav) {
+        if (!isInState(loginNav)) {
+            mLoginNavStack.push(loginNav);
+        }
+    }
+
+    public void goBack() {
+        if (mLoginNavStack.isEmpty()) {
+            throw new RuntimeException("Navigation stack is empty! Can't go back.");
+        }
+
+        mLoginNavStack.pop();
+    }
+
+    public boolean isNavStackEmpty() {
+        return mLoginNavStack.isEmpty();
+    }
+
+    // available for testing purposes. Don't use otherwise
+    public void force(Class<? extends LoginNav> loginNav) {
+        mLoginNavStack.push(loginNav);
+    }
+
+    private class PrologueHandler implements LoginNav.Prologue {
+        @Override
+        public void doStartLogin() {
+            ensureState(LoginNav.Prologue.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.showEmailLoginScreen();
+        }
+
+        @Override
+        public void doStartSignup() {
+            ensureState(LoginNav.Prologue.class);
+            gotoState(LoginNav.Prologue.class);
+
+            mLoginNavHandler.toast("Signup is not implemented yet");
+        }
+    }
+
+    private class InputEmailHandler implements LoginNav.InputEmail {
+        @Override
+        public void gotEmail(String email) {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Input email is not implemented yet. Input email: " + email);
+        }
+
+        @Override
+        public void loginViaUsernamePassword() {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Fall back to username/password is not implemented yet.");
+        }
+
+        @Override
+        public void help() {
+            ensureState(LoginNav.InputEmail.class);
+            gotoState(LoginNav.InputEmail.class);
+
+            mLoginNavHandler.toast("Help is not implemented yet.");
+        }
+    }
+
+    private class InputSiteAddressHandler implements LoginNav.InputSiteAddress {
+        @Override
+        public void gotSiteAddress(String siteAddress) {
+            ensureState(LoginNav.InputSiteAddress.class);
+            gotoState(LoginNav.InputSiteAddress.class);
```

Comment:
> so I'd imagine we'd never want to pass the same state. Am I mistaken about that?

You're not mistaken at all. By the time `gotSiteAddress` (and its counterparts in other screens) is called, the screen/Fragment has already decided that a screen change needs to occur and will call its parent Activity to make it so.

In this case, the transition out of the Site Address screen is not implemented, that's why the `goto` state is the same as the `ensure` line. `doStartLogin` is the only place the 2 differ because that's the only transition so far implemented; the rest will come in other PRs. Notice that all these are also tested in `LoginLogicTest` where the state non-change is tested (`inputSiteAddressGotSiteAddressTest` test case). I should have added a line comment to all these no-state-changes to make it clear.